### PR TITLE
Add SSD network

### PR DESCRIPTION
**Pull request by joaqo**
Achieves 71mAP on PASCAL VOC2007 test set when trained on union of VOC2007 and VOC2012 trainval. Paper quotes 74 mAP for this metric.

The difference could be explained by slight differences in our implementations of data augmentation and our usage of batch size 1 instead of 32. When, testing other implementations of SSD, batch size 1 vs 32 has shown to make a significant difference on the final mAP.

Could use a general refactor and needs tests, but that was determined not to be a top priority atm.

### CODE REVIEW

File: luminoth/models/ssd/ssd_feature_extractor.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
+import tensorflow as tf
```
Comment (nagitsu) [inline luminoth/models/ssd/ssd_feature_extractor.py]: Remove the `ssd_` at the start of the file, it's already prefixed by the package.

File: luminoth/models/ssd/ssd_feature_extractor.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        super(SSDFeatureExtractor, self).__init__(config, name=name, **kwargs)
+        if self._architecture not in VALID_SSD_ARCHITECTURES:
+            raise ValueError('Invalid architecture "{}"'.format(
+                self._architecture
+            ))
+        self.parent_name = parent_name
+        self.activation_fn = tf.nn.relu
+
+    def _init_vgg16_extra_layers(self):
+        # TODO: Try Xavier initializer
```
Comment (nagitsu) [inline luminoth/models/ssd/ssd_feature_extractor.py]: Remove this comment and move to TODO list?

File: luminoth/models/ssd/ssd_target.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
+import tensorflow as tf
+import sonnet as snt
+
+from luminoth.utils.bbox_transform_tf import encode
+from luminoth.utils.bbox_overlap import bbox_overlap_tf
+
+
+class SSDTarget(snt.AbstractModule):
+    """TODO
```
Comment (nagitsu) [inline luminoth/models/ssd/ssd_target.py]: :thinking:

File: luminoth/models/ssd/ssd_utils.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    x_max = x_max * new_width
+    y_max = y_max * new_height
+
+    # Concat points and label to return a [num_bboxes, 4] tensor.
+    return np.stack([x_min, y_min, x_max, y_max], axis=1)
+
+
+def generate_anchors_reference(ratios, scales, num_anchors, feature_map_shape):
+    """
+    For each ratio we will get an anchor TODO
```
Comment (nagitsu) [inline luminoth/models/ssd/ssd_utils.py]: :thinking:

File: luminoth/models/base/base_network.py [hunk 5]
```diff
@@ -183,11 +183,20 @@ def load_weights(self):
...
             # running locally, or to the job bucket if running in Google Cloud.
+
+            # TODO: Shouldn't _config['weights'] be called weights_path or
+            # something similar?
             self._config['weights'] = get_checkpoint_file(self._architecture)
 
-        module_variables = snt.get_variables_in_module(
-            self, tf.GraphKeys.MODEL_VARIABLES
-        )
+        if self.pretrained_weights_scope:
```
Comment (vierja) [inline luminoth/models/base/base_network.py]: Add comment explaining why this is needed.

File: luminoth/utils/predicting.py [hunk 6]
```diff
@@ -71,7 +71,12 @@ def __init__(self, config):
                 )
                 self.session.run(init_op)
 
-            if config.model.network.with_rcnn:
+            if config.model.type == 'ssd':
+                cls_prediction = pred_dict['classification_prediction']
+                objects_tf = cls_prediction['objects']
+                objects_labels_tf = cls_prediction['labels']
+                objects_labels_prob_tf = cls_prediction['probs']
+            elif config.model.network.get('with_rcnn', False):
```
Comment (nagitsu) [inline luminoth/utils/predicting.py]: Nested `if`: first model, then `with_rcnn` check.

File: luminoth/models/ssd/base_config.yml [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+  image_vis: debug
+
+eval:
+  # Image visualization mode, options = train, eval, debug, (empty). Default=(empty)
+  image_vis: eval
+
+dataset:
+  type: tfrecord
+  # From which directory to read the dataset
+  dir: /data/datasets/pascal-2007-2012/
```
Comment (vierja) [inline luminoth/models/ssd/base_config.yml]: Should be relative (same as Faster's or removed)

File: luminoth/models/ssd/base_config.yml [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+  seed:
+  # Training batch size for images.
+  batch_size: 1
+  # Directory in which model checkpoints & summaries (for Tensorboard) will be saved
+  job_dir: jobs/
+  # Enables TensorFlow debug mode, which stops and lets you analyze Tensors
+  # after each Session.run
+  tf_debug: False
+  # Run name is used for better checkpoint and log hierarchy
+  run_name: ssd
```
Comment (vierja) [inline luminoth/models/ssd/base_config.yml]: I would prefer keeping this default empty.

File: luminoth/models/ssd/ssd.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
+import numpy as np
+import sonnet as snt
+import tensorflow as tf
+
+from sonnet.python.modules.conv import Conv2D
+from luminoth.models.ssd.ssd_feature_extractor import SSDFeatureExtractor
```
Comment (vierja) [inline luminoth/models/ssd/ssd.py]: Maybe sort and remove empty lines?

File: luminoth/models/ssd/ssd.py [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+                num_anchors * (self._num_classes + 1), [3, 3],
+                name=feat_map_name + '_classes_conv',
+            )(feat_map)
+            class_scores_flattened = tf.reshape(
+                class_scores_layer, [-1, self._num_classes + 1]
+            )
+            class_scores_list.append(class_scores_flattened)
+        bbox_offsets = tf.concat(bbox_offsets_list, axis=0)
+        class_scores = tf.concat(class_scores_list, axis=0)
+        class_probabilities = tf.nn.softmax(class_scores)
```
Comment (vierja) [inline luminoth/models/ssd/ssd.py]: I would add the `axis` kwarg to make it explicit.

File: luminoth/models/ssd/ssd.py [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+            #       them to their feature map size.
+            feat_map_shape = feat_map.shape.as_list()[1:3]
+            scaled_bboxes = adjust_bboxes(
+                raw_anchors_per_featmap[feat_map_name], feat_map_shape[0],
+                feat_map_shape[1], self.image_shape[0], self.image_shape[1]
+            )
+            clipped_bboxes = clip_boxes(scaled_bboxes, self.image_shape)
+            anchors_list.append(clipped_bboxes)
+        anchors = np.concatenate(anchors_list, axis=0)
+        # TODO: They were using float64, is all this precision necesary?
```
Comment (vierja) [inline luminoth/models/ssd/ssd.py]: It's not necessary. Replace with 32.

File: luminoth/models/ssd/ssd.py [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+from luminoth.utils.bbox_transform import clip_boxes
+
+
+class SSD(snt.AbstractModule):
+    """SSD: Single Shot MultiBox Detector
+    """
+
+    def __init__(self, config, name='ssd'):
+        super(SSD, self).__init__(name=name)
+        self._config = config
```
Comment (vierja) [inline luminoth/models/ssd/ssd.py]: Only config.model.* variables are used. Maybe save that as _config instead?

File: luminoth/models/ssd/ssd_target.py [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+            anchors_gt_boxes,
+            variances=self._variances
+        )
+
+        # We unmap targets to anchor_labels (containing the length of
+        # anchors)
+        bbox_targets = tf.scatter_nd(
+            indices=anchors_with_target_idx,
+            updates=bbox_targets,
+            shape=tf.cast(tf.shape(all_anchors), tf.int64)
```
Comment (vierja) [inline luminoth/models/ssd/ssd_target.py]: tf.int or 32

File: luminoth/models/ssd/ssd_target.py [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+
+    def _build(self, probs, all_anchors, gt_boxes, im_shape):
+        """
+        Args:
+            all_anchors: A Tensor with anchors for all of SSD's features.
+                The shape of the Tensor is (num_anchors, 4).
+            gt_boxes: A Tensor with the ground truth boxes for the image.
+                The shape of the Tensor is (num_gt, 5), having the truth label
+                as the last value for each box.
+            im_shape: (1, height, width, )
```
Comment (vierja) [inline luminoth/models/ssd/ssd_target.py]: im_shape is not used.

File: luminoth/models/ssd/ssd.py [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
+            gt_boxes = tf.cast(gt_boxes, tf.float32)
+
+            # Generate targets
+            target_creator = SSDTarget(
+                self._num_classes, anchors.shape[0], self._config.model.target,
+                self._config.model.variances
+            )
+            class_targets, bbox_offsets_targets = target_creator(
+                class_probabilities, anchors, gt_boxes,
+                tf.cast(tf.shape(image), tf.float32)
```
Comment (vierja) [inline luminoth/models/ssd/ssd.py]: ~~Unnecessary cast~~ unused value

File: luminoth/models/ssd/ssd_target.py [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+                 name='ssd_target'):
+        """
+        Args:
+            num_classes: Number of possible classes.
+            config: Configuration object for RCNNTarget.
+        """
+        super(SSDTarget, self).__init__(name=name)
+        self._num_classes = num_classes
+        self._foreground_fraction = config.hard_negative_ratio
+        self._num_anchors = tf.cast(num_anchors, tf.int32)
```
Comment (vierja) [inline luminoth/models/ssd/ssd_target.py]: Unused

File: luminoth/models/ssd/ssd_target.py [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+        self._foreground_threshold = config.foreground_threshold
+        self._background_threshold_high = config.background_threshold_high
+        self._background_threshold_low = config.background_threshold_low
```
Comment (vierja) [inline luminoth/models/ssd/ssd_target.py]: Add comment that is not used but kept for compatibility.

File: luminoth/models/ssd/ssd.py [hunk 18]
```diff
@@ -0,0 +1,338 @@
...
+                tf.cast(tf.shape(image)[1:3], tf.float32)
+            )
+            prediction_dict['classification_prediction'] = proposals
+
+        # Add some non essential metrics for debugging
+        if self._debug:
+            prediction_dict['all_anchors'] = anchors
+            prediction_dict['cls_prob'] = class_probabilities
+            if gt_boxes:
+                # TODO does it make sense to return this when its an input
```
Comment (vierja) [inline luminoth/models/ssd/ssd.py]: It was used for image_vis and was called `gt_bboxes` not `gt_boxes`, so no, it doesn't make sense to keep it.

File: luminoth/models/ssd/ssd_feature_extractor.py [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+        """
+        # TODO: Is there a better way to manage scoping in these cases?
+        scope = self.module_name
+        if self.parent_name:
+            scope = self.parent_name + '/' + scope
+
+        base_net_endpoints = super(SSDFeatureExtractor, self)._build(
+            inputs, is_training=is_training)['end_points']
+
+        if self.vgg_type:
```
Comment (vierja) [inline luminoth/models/ssd/ssd_feature_extractor.py]: vgg_type includes VGG19.

File: luminoth/models/ssd/ssd_feature_extractor.py [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+                    'gamma', shape=params_shape, dtype=dtype,
+                    initializer=init_ops.ones_initializer()
+                )
+                vgg_conv4_3_norm = tf.multiply(vgg_conv4_3_norm, scale)
+            tf.add_to_collection('FEATURE_MAPS', vgg_conv4_3_norm)
+
+            # Extra layers for vgg16 as detailed in paper
+            self._init_vgg16_extra_layers()
+            with tf.variable_scope('extra_feature_layers'):
+                # from IPython import embed; embed(display_banner=False)
```
Comment (vierja) [inline luminoth/models/ssd/ssd_feature_extractor.py]: Remove.

File: luminoth/models/ssd/ssd_feature_extractor.py [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+                tf.add_to_collection('FEATURE_MAPS', net)
+                net = self.conv11_1(net)
+                net = self.activation_fn(net)
+                net = self.conv11_2(net)
+                net = self.activation_fn(net)
+                tf.add_to_collection('FEATURE_MAPS', net)
+
+            # This parameter determines onto which variables we try to load the
+            # pretrained weights
+            self.pretrained_weights_scope = 'ssd/ssd_feature_extractor/vgg_16'
```
Comment (vierja) [inline luminoth/models/ssd/ssd_feature_extractor.py]: This should use `module_name` and `architecture`.

File: luminoth/models/ssd/ssd_proposal.py [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+    and regression (adjusting the bounding box), we return a list of proposals
+    with assigned class.
+
+    In the process it tries to remove duplicated suggestions by applying non
+    maximum suppresion (NMS).
+
+    We apply NMS because the way object detectors are usually scored is by
+    treating duplicated detections (multiple detections that overlap the same
+    ground truth value) as false positive. It is resonable to assume that there
+    may exist such case that applying NMS is completly unnecesary.
```
Comment (vierja) [inline luminoth/models/ssd/ssd_proposal.py]: completly

File: luminoth/models/ssd/ssd_proposal.py [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+            raw_proposals = decode(anchors, class_loc_pred, self._variances)
+            # Clip boxes to image.
+            clipped_proposals = clip_boxes(raw_proposals, im_shape)
+
+            # Filter proposals that have an non-valid area.
+            (x_min, y_min, x_max, y_max) = tf.unstack(
+                clipped_proposals, axis=1)
+            proposal_filter = tf.greater(
+                tf.maximum(x_max - x_min, .0) * tf.maximum(y_max - y_min, .0),
+                .0
```
Comment (vierja) [inline luminoth/models/ssd/ssd_proposal.py]: `0.` 🙊

File: luminoth/models/ssd/ssd_target.py [hunk 24]
```diff
@@ -0,0 +1,194 @@
...
+    def __init__(self, num_classes, num_anchors, config, variances, seed=None,
+                 name='ssd_target'):
+        """
+        Args:
+            num_classes: Number of possible classes.
+            config: Configuration object for RCNNTarget.
+        """
+        super(SSDTarget, self).__init__(name=name)
+        self._num_classes = num_classes
+        self._foreground_fraction = config.hard_negative_ratio
```
Comment (vierja) [inline luminoth/models/ssd/ssd_target.py]: This is confusing. Why the different names?

File: luminoth/utils/image.py [hunk 25]
```diff
@@ -531,3 +564,59 @@ def random_distortion(image, bboxes=None, brightness=None, contrast=None,
...
+    pad_top = tf.random_uniform([1], minval=0,
+                                maxval=new_height-height, seed=seed)
+    pad_bottom = new_height - height - pad_top
+    paddings = tf.stack([tf.concat([pad_top, pad_bottom], axis=0),
+                         tf.concat([pad_left, pad_right], axis=0),
+                         tf.constant([0., 0.])])
+    expanded_image = tf.pad(image, tf.to_int32(paddings), constant_values=fill)
+
+    # Adjust bboxes
+    # bboxes = tf.Print(bboxes, [bboxes], summarize=300)
```
Comment (vierja) [inline luminoth/utils/image.py]: Remove comment.

File: luminoth/utils/image.py [hunk 26]
```diff
@@ -531,3 +564,59 @@ def random_distortion(image, bboxes=None, brightness=None, contrast=None,
...
+        bboxes: Optional Tensor with bounding boxes with shape (num_bboxes, 5).
+            where we have (x_min, y_min, x_max, y_max, label) for each one.
+
+    Returns:
+        Dictionary containing:
+            image: Tensor with zoomed out image.
+            bboxes: Tensor with zoomed out bounding boxes with shape
+                (num_bboxes, 5).
+    """
+    # image = tf.Print(image, [image], summarize=500, message='lol')
```
Comment (vierja) [inline luminoth/utils/image.py]: Remove comment.

File: luminoth/utils/image.py [hunk 27]
```diff
@@ -531,3 +564,59 @@ def random_distortion(image, bboxes=None, brightness=None, contrast=None,
...
+    expanded_image = tf.pad(image, tf.to_int32(paddings), constant_values=fill)
+
+    # Adjust bboxes
+    # bboxes = tf.Print(bboxes, [bboxes], summarize=300)
+    shift_bboxes_by = tf.concat([pad_left, pad_top, pad_left, pad_top], axis=0)
+    bbox_labels = tf.reshape(bboxes[:, 4], (-1, 1))
+    bbox_adjusted_coords = bboxes[:, :4] + tf.to_int32(shift_bboxes_by)
+    bbox_adjusted = tf.concat([bbox_adjusted_coords, bbox_labels], axis=1)
+
+    # TODO: use mean instead of 0 for fill
```
Comment (vierja) [inline luminoth/utils/image.py]: Move to docstring.

File: luminoth/utils/predicting.py [hunk 28]
```diff
@@ -109,8 +114,12 @@ def predict_image(self, image):
         if self.class_labels is not None:
             labels = [self.class_labels[label] for label in labels]
 
-        # Scale objects to original image dimensions.
-        objects /= scale_factor
+        # Scale objects to original image dimensions
+        if isinstance(scale_factor, list):
+            objects /= [scale_factor[1], scale_factor[0],
```
Comment (vierja) [inline luminoth/utils/predicting.py]: Add comment on what scale factor is when it is not a single value. (Also, maybe use tuple instead of list?)

File: luminoth/models/ssd/ssd_target.py (hunk 13 above)
Comment (joaqo) [reply luminoth/models/ssd/ssd_target.py]: We use int64 because `indices` is int64, and this is because `indices` comes from an `tf.where` with no x and y arguments, which is hardcoded to return int64 by tensorflow.
